### PR TITLE
Features/002 Parser

### PR DESCRIPTION
**Pull request by geokala**
Includes contrib tools scripts to assist testing, but these should be factored out later.

### CODE REVIEW

**Comment by alistair-broomhead** [comment]
Going to take a quick look in my lunch today, should have completed review by the next project nights ;)

File: contrib_tools/prepare [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#! /usr/bin/env bash
```
Comment (alistair-broomhead) [inline contrib_tools/prepare:1]: It would be good if these contrib tools could be split into a separate PR, as there's rather a lot going on here.

File: src/romaine/core.py [hunk 2]
```diff
@@ -14,6 +15,7 @@ def __init__(self):
             Initialise Romaine core.
         """
         self.steps = {}
+        self.parser = Parser()
```
Comment (alistair-broomhead) [inline src/romaine/core.py]: I'm not entirely sure why the parser has to be attached as an attribute of the core - is it not something the core uses to get the parser's result?

File: src/romaine/parser/__init__.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+from romaine.parser import (
+    simple,
+    multiline,
+    step,
+    section,
+    feature,
+    )
```
Comment (alistair-broomhead) [inline src/romaine/parser/__init__.py]: (minor) These closing parens are overindented (should line up with start of statement)

File: src/romaine/parser/__init__.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+from romaine.parser.exceptions import (  # noqa
+    UnclosedPythonishString,
+    MalformedTableError,
+    )
+
+
+class Parser(object):
+    """
+        Gherkin parser for romaine core.
```
Comment (alistair-broomhead) [inline src/romaine/parser/__init__.py:16]: (minor) These docstrings are overindented (should line up with start of triple-quotes)

**Comment by geokala** [comment]
Excellent. I started hacking on the json schemas- they're not 100%
necessary for 1.0, but I needed to relax a bit and json schemas were
easier. _smiles_

S

File: src/romaine/parser/simple.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        """
+            Takes a line and determines whether it is pure whitespace.
+
+            Keyword arguments:
+            line - Line to test.
+
+            Returns:
+            True if the line is purely whitespace. False otherwise.
+        """
+        line = line.strip()
```
Comment (alistair-broomhead) [inline src/romaine/parser/simple.py]: This is fine, but what's the difference in behaviour between this and https://docs.python.org/2/library/stdtypes.html#str.isspace ?

**Comment by alistair-broomhead** [comment]
I still need to look through the tests (and the contrib tools, but I'd prefer these were seperated into another PR), but the parser code looks good - nice work!

File: src/romaine/core.py (hunk 2 above)
Comment (geokala) [inline src/romaine/core.py]: It is, but it feels like the core should have all of the components it used (well, feels like that to me anyway). Much the same as we have the ability to locate features in the core, this then gives the ability to do something with those features without leaving the core.

Happy to consider a different approach if we've got a consistent paradigm we're trying to stick to with the core.

File: src/romaine/parser/__init__.py (hunk 3 above)
Comment (geokala) [inline src/romaine/parser/__init__.py]: Yeah, I was trying to do all of these in the style where you do closing parens on the same level of indentation as the thing they're closing. I'm not sure I was consistent on that though. Inherited it from a friend and I actually think it looks nicer but it's so infrequently done that I'll sort it now.

Un(?)fortunately, flake8 doesn't complain about it, so I'll have to search for it by hand.

File: src/romaine/parser/__init__.py (hunk 4 above)
Comment (geokala) [inline src/romaine/parser/__init__.py:16]: Ergh- that style always looks horrible to me. I much prefer doing the docstrings with the same indentation approach used everywhere else (given that I newline it immediately, because it looks ugly to have text immediately following the triple quotes).

How attached are you to this element of style? I can go through and change it, but currently that style is used everywhere (including in the core).

File: src/romaine/parser/simple.py (hunk 5 above)
Comment (geokala) [inline src/romaine/parser/simple.py]: "".isspace() => False

I can either modify to:
return len(line) == 0 or line.isspace()

or I can leave as is. No strong feelings either way for me.

**Comment by geokala** [comment]
Contrib tools have been pulled out and will be pushed back up as an unaffiliated PR.

I had to make flake8 ignore the new E402 (import not at top of file) because that complains on our tests, and on anything that requires setting anything before an import. I suppose we could allow it again and noqa it in the tests/common.py

File: src/romaine/parser/__init__.py (hunk 4 above)
Comment (alistair-broomhead) [inline src/romaine/parser/__init__.py:16]: I think we both have the same reaction to each other's preferred convention ;)

Obviously I'd prefer it my way, but I can do that as a separate PR once all of this is in.

File: src/romaine/parser/simple.py (hunk 5 above)
Comment (alistair-broomhead) [inline src/romaine/parser/simple.py]: Yep the above proposal is good - it might be good to mention this difference in the docstring, but it should be reasonably obvious from the proposed implementation.

File: src/romaine/core.py (hunk 2 above)
Comment (alistair-broomhead) [inline src/romaine/core.py]: Maybe it's better to bind a label to the `Parser` class into `Core`'s namespace? That way the class, and its instances, have everything required, without having an instance of the `Parser` hanging around on every copy?

File: tests/test_parser_steps.py [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+        Test steps gherkin parser functionality of romaine's core.
+    """
+    def setUp(self):
+        """
+            Prepare the environment for testing.
+        """
+        # We're doing a lot of long tests- don't limit the diff output length
+        self.maxDiff = None
+
+    def tearDown(self):
```
Comment (alistair-broomhead) [inline tests/test_parser_steps.py:16]: Can this not be left out in the case that we're not doing any teardown?

File: tests/test_parser_steps.py [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+
+    def test_getting_raw_step(self):
+        """
+            Check we can get a step with no surrounding data.
+        """
+        # Given I have Romaine core's parser
+        parser = common.get_romaine_parser()
+
+        # When I call get_step on a list containing
+        # """[
```
Comment (alistair-broomhead) [inline tests/test_parser_steps.py:60]: Not an issue at this point, but I might be tempted to represent this in gherkin as 

```
When I call get_step on a list containing:
  | item                                  |
  | When I call get_step on an empty list |
```

File: tests/test_parser_steps.py [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+    def test_get_step_empty(self):
+        """
+            Confirm that we get nothing when we don't start with a step.
+        """
+        # Given I have Romaine core's parser
+        parser = common.get_romaine_parser()
+
+        # When I call get_step on a list containing:
+        # """[
+        #     'This is not a real step.',
```
Comment (alistair-broomhead) [inline tests/test_parser_steps.py:136]: Gah - we might want to relax this in the future, as other frameworks are not so strict

File: tests/test_parser_simple.py [hunk 9]
```diff
@@ -0,0 +1,794 @@
+from tests import common
+import unittest
+
+
+class TestSimpleParser(unittest.TestCase):
+    """
+        Test single line gherkin parser functionality of romaine's core.
+    """
+    def setUp(self):
```
Comment (alistair-broomhead) [inline tests/test_parser_simple.py:9]: Not sure these are needed (I'll not mention it for the other files)

File: tests/test_parser_steps.py [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+            '"""',
+            '',
+            '# Test second step',
+            '',
+            'Then I am happy to see this step',
+            'No more diagonal lines',
+        ]
+        result = parser.step.get_steps(input_data)
+
+        # Then the result is two steps: type=When with
```
Comment (alistair-broomhead) [inline tests/test_parser_steps.py:563]: I might be tempted to do this as

```
Then the result is two steps:
  | type | text | multiline_arg_type | multiline_arg_value | trailing_whitespace |
....
```

File: tests/common.py [hunk 11]
```diff
@@ -7,4 +7,9 @@
 sys.path.append(package_directory)
 
 # Module to be tested
-import romaine  # noqa
+import romaine
+
+
+# Utility for getting an initialised parser.
+def get_romaine_parser():
+    return romaine.Core().parser
```
Comment (alistair-broomhead) [inline tests/common.py]: if we change the parser instantiation thing this could be `return romaine.Core.Parser()`

File: src/romaine/core.py (hunk 2 above)
Comment (geokala) [inline src/romaine/core.py]: Sounds sensible, sorting that out now.

File: src/romaine/parser/__init__.py (hunk 4 above)
Comment (geokala) [inline src/romaine/parser/__init__.py:16]: Well, if we put in the docs for the project which documentation style we prefer (there's an issue open for that anyway) then I'm happy as long as we're consistent. Even if we're consistently ugly. _smiles_

File: src/romaine/parser/simple.py (hunk 5 above)
Comment (geokala) [inline src/romaine/parser/simple.py]: I'll add a comment to the line- the implementation detail probably doesn't need to live in the docstring.

File: tests/test_parser_steps.py (hunk 6 above)
Comment (geokala) [inline tests/test_parser_steps.py:16]: It could, I just quite like having it there to be explicit and provide a nice framework for new tests.

File: tests/test_parser_steps.py (hunk 7 above)
Comment (geokala) [inline tests/test_parser_steps.py:60]: Yeah, let's sort these all out properly as part of implementing the tests of itself?

File: tests/test_parser_steps.py (hunk 8 above)
Comment (geokala) [inline tests/test_parser_steps.py:136]: How so? This is only called where it should get steps (and in those cases I can't see a good reason for it to find steps when there aren't steps).

Probably best if you can come up with a test that shows when it's a good idea for this not to happen so that we can code against it?

File: tests/test_parser_steps.py (hunk 10 above)
Comment (geokala) [inline tests/test_parser_steps.py:563]: See response regarding the other BDD style test comments.

I think these all need reworking, and wrapping around a bunch of functions which we could then use in both the 'traditional' tests and as the backend for the BDD ones.

File: tests/test_parser_simple.py (hunk 9 above)
Comment (geokala) [inline tests/test_parser_simple.py:9]: See other response. I'm happy to remove these if you prefer. I think they're currently in all of the tests, so they should at least be consistent, but if we'd prefer for them to consistently be absent I can rip them out.

File: tests/test_parser_steps.py (hunk 8 above)
Comment (alistair-broomhead) [inline tests/test_parser_steps.py:136]: Sure - shall we open an issue for it, milestone 1.1?

Comment (geokala) [inline tests/test_parser_steps.py:136]: Sounds good- perhaps just one to create a bunch more full blueprint tests though? (Given that we have... two?)

Comment (alistair-broomhead) [inline tests/test_parser_steps.py:136]: Here's an example form lettuce: http://lettuce.it/intro/wtf.html#lettuce-outlined

File: tests/test_parser_steps.py (hunk 7 above)
Comment (alistair-broomhead) [inline tests/test_parser_steps.py:60]: definitely - just wanted to get in some early communication ;)

**Comment by alistair-broomhead** [comment]
I think I'd rather not have empty methods, as it's just more useless text that doesn't do anything while you're looking for functionality. This isn't the end of the world though, we won't be using unittest once we reach 1.0 anyway.

Barring the discussed style issues, which we can sort when we create a style guide, I think this is good to go, what do you think? Happy to merge this, or let you do it - if we can get this merged tonight, I might work on some of the other tickets tomorrow in our Skimlinks innovation day before the meetup.

**Comment by geokala** [comment]
That's fair, and yes- I should've thought about that when I was building it as I'm not overly keen on such things.

Regarding using unittest, I thought the plan was to keep the unittest bit for the absolute core that is required in order for the BDD style tests to actually work. Anything beyond that should be BDD only, meaning that we don't suffer great pain when we break the core and have no way to test it that doesn't rely on the core.

And yes, let's leave style for the style guide for the moment so that we can get the basics of this up and running. I'll merge now.
